README.md
# PHP Article extractor

This is a web article parsing and language detection library for PHP. This library reads the article content from a web page, removing all HTML and providing just the raw text, suitable for text to speech or machine learning processes.

For a project I have developed, I found many existing open source solutions good starting points, but each had unique failures. This library aggregates three different approaches into a single solution while adding the additional functionality of language detection.

## How To Use

This library is distributed via packagist.org, so you can use composer to retrieve the dependency

```
composer require crscheid/php-article-extractor
```

### Calling via URL

This library will attempt to retrieve the HTML for you. You need simply to create an ArticleExtractor class and call the `parseURL` function on it, passing in the URL desired.

```php
use Cscheide\ArticleExtractor\ArticleExtractor;

$extractor = new ArticleExtractor();

$response = $extractor->processURL("https://www.fastcompany.com/3067246/innovation-agents/the-unexpected-design-challenge-behind-slacks-new-threaded-conversations");
var_dump($response);
```

The function `processURL` returns an array containing the title, text, and meta data associated with the request. If the text is `null` then this indicates a failed parsing. Below should be the output of the above code.

The field `result_url` will be different if the library followed redirects. This field represents the final page actually retrieved after redirects.

```
array(5) {
  ["parse_method"]=>
  string(11) "readability"
  ["title"]=>
  string(72) "The Unexpected Design Challenge Behind Slack’s New Threaded Conversations"
  ["text"]=>
  string(8013) "At first blush, threaded conversations sound like one of the most thoroughly mundane features a messaging app could introduce.After all, the idea of neatly bundling up a specific message and its replies in ..."
  ["language_method"]=>
  string(7) "service"
  ["language"]=>
  string(2) "en"
  ["result_url"]=>
  string(126) "https://www.fastcompany.com/3067246/innovation-agents/the-unexpected-design-challenge-behind-slacks-new-threaded-conversations"

}
```

### Calling with HTML

If you already have HTML, you can use the `parseHTML` function and use your HTML processed through the same logic.

```php
use Cscheide\ArticleExtractor\ArticleExtractor;

$extractor = new ArticleExtractor();
$myHTML = <load from some source>;

$response = $extractor->processHTML($myHTML);
var_dump($response);
```

The function `parseHTML` returns an array containing the title, text, and meta data associated with the request. If the text is `null` then this indicates a failed parsing. Below should be the output of the above code.

The field `result_url` will not be included in this case since we are not attempting to get the HTML during the process call.

```
array(5) {
  ["parse_method"]=>
  string(11) "readability"
  ["title"]=>
  string(72) "The Unexpected Design Challenge Behind Slack’s New Threaded Conversations"
  ["text"]=>
  string(8013) "At first blush, threaded conversations sound like one of the most thoroughly mundane features a messaging app could introduce.After all, the idea of neatly bundling up a specific message and its replies in ..."
  ["language_method"]=>
  string(7) "service"
  ["language"]=>
  string(2) "en"
}
```


You can also create the `ArticleExtractor` class by passing in a key for the language detection service as well as a custom User-Agent string. See more information below.


## Options

### Language Detection Methods

Language detection is handled by either looking for language specifiers within the HTML meta data or by utilizing the [Detect Language](http://detectlanguage.com/) service.

If it is possible to detect the language of the article, the language code in [ISO 639-1](http://www.loc.gov/standards/iso639-2/php/code_list.php) format as well as the detection method are returned in the fields `language` and `language_method` respectively. The `language_method` field, if found successfully, may be either `html` or `service`.

If language detection fails or is not available, both of these fields will be returned as null.

[Detect Language](http://detectlanguage.com/) requires the use of an API KEY which you can sign up for. However, you can also use this library without it. If the HTML meta data do not contain information about the language of the article, then `language` and `language_method` will be returned as null values.

To utilize this library utilizing the language detection service, create the `ArticleExtractor` object by passing in your API KEY for [Detect Language](http://detectlanguage.com/).

```php
use Cscheide\ArticleExtractor\ArticleExtractor;

$extractor = new ArticleExtractor('your api key');
```

### Setting User Agent

It is possible to set the user-agent for outgoing requests. To do so pass the desired user agent string to the constructor as follows:

```php
use Cscheide\ArticleExtractor\ArticleExtractor;

$myUserAgent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.116 Safari/537.36";
$extractor = new ArticleExtractor(null, $myUserAgent);
```

### Force Reading Method

It is possible to force the method by which the reading is attempted, either with Readability, Goose, or Goose with our custom processing. This can come in handy where Readability or Goose have particular issues with particular websites.

To force the method, simply provide a third argument to the constructor as such. The four valid methods are `readability`, `goose`, `goosecustom`, or `custom`.

```php
$extractor = new ArticleExtractor(null, null, "goose");
```


## Output Format

As of version 1.0, the output format has been altered to provide newline breaks for headings. This is important especially for natural language processing applications in determining sentence boundaries. If this behavior is not desired, simply strip out the additional newlines where needed.

This change was made due the fact that when header and paragraph HTML elements are simply stripped out, there often occurs issues where there is no separation between the heading and the proceeding sentence.

**Example of Output Format for Text Field**

```
\n
A database containing 250 million Microsoft customer records has been found unsecured and online\n
NurPhoto via Getty Images\n
A new report reveals that 250 million Microsoft customer records, spanning 14 years, have been exposed online without password protection.\n
Microsoft has been in the news for, mostly, the wrong reasons recently. There is the Internet Explorer zero-day vulnerability that Microsoft hasn't issued a patch for, despite it being actively exploited. That came just days after the U.S. Government issued a critical Windows 10 update now alert concerning the "extraordinarily serious" curveball crypto vulnerability. Now a newly published report, has revealed that 250 million Microsoft customer records, spanning an incredible 14 years in all, have been exposed online in a database with no password protection.\n
What Microsoft customer records were exposed online, and where did they come from?\n
```




## Running tests

Unit tests are included in this distribution and can be run utilizing PHPUnit after installing dependencies. The recommended approach is to use Docker for this purpose, so you then don't even need to have dependencies installed on your system.

> Note: Please set the environment variable `DETECT_LANGUAGE_KEY` with your [Detect Language](http://detectlanguage.com/) key in order for language detection in unit tests to work properly.


### Installing Dependencies

This will use the composer docker image to download the requirements. Note the use of the `--ignore-platform-reqs` since some of our dependencies do not yet support PHP 8. Ensure that whatever platform you are running composer on is on version 7.2 or higher of PHP.

```
docker run --rm --interactive --tty --volume $PWD:/app composer --ignore-platform-reqs install
```

### Running Unit Tests

This runs the phpunit dependency that we downloaded within the php 7.4 command line environment.

```
docker run -v $(pwd):/app -w /app -e DETECT_LANGUAGE_KEY=<yourapikey> --rm php:7.4-cli ./vendor/phpunit/phpunit/phpunit
```


CHANGELOG.md
# ChangeLog

## Version 2.5.3

- Locks composer dependencies to specific versions
  - PHP to ~7.2
  - scotteh/php-goose to ^1.1


## Version 2.5.2

- Updated `parseURLViaReadability` function to include Accept header

## Version 2.5.1

- Migrates `andreskrey/readability.php` to [fivefilters/readability.php](https://packagist.org/packages/fivefilters/readability.php)


## Version 2.5

- Fixes missing $html bug in ArticleExtractor.

## Version 2.3

- Added ability to pass in HTML and process via the `processHTML` method
- Revised documentation with updates and fixes
- Updated test cases

## Version 2.2

- Added ability to override/force the reading method.

## Version 2.1

- Added handling of common Google referral URLs
- Added 'result_url' to the return structure to inform the caller what the resultant URL was after redirects

## Version 2.0.1

- Turned off debugging left on by mistake

## Version 2.0

- Added ability to manually set User-Agent, fixing many readability issues
- Updated redirect detection logic to more accurately read HTTP headers.
- Updated dependencies
  - Updated PHPUnit to ^8.0
  - Updated andreskrey/readability.php to ^2.1.0
- Updated PHP dependency to ^7.2

## Version 1.0.1

- Fixed minor issue with `parse_url` check.

## Version 1.0

- Updated to modify the approach for cleaning HTML tags and dealing with newlines.
- Updated README.md to outline the new text format.
- Closes issue #25

## Version 0.9

- Updated to include cleaning up of article text.

## Version 0.8.5

- Updated redirect checking logic to include ports


## Version 0.8.4

- Resolved 301 redirects to incomplete URL

## Version 0.8.3

- Closes issue #23 related to 301 redirects when scheme and host is not present.

## Version 0.8

- Added [andreskrey/readability.php](https://github.com/andreskrey/readability.php) library as the default method of article parsing, using prior methods as a backup.
  - This closes multiple issues related to article reading including #6, #7, #8, #17, #18
- Changed the call to parse a URL from `getArticleText to `processURL`
- Added README.md
- Upgraded to PHPUnit 6.x for testing


## Version 0.7

- Started CHANGELOG
- Moved PHP DOM Parser dependency to [thesoftwarefanatics/php-html-parser](https://github.com/thesoftwarefanatics/php-html-parser) for support of PHP 7.2+


tests/ExtractorTest.php
<?php

use Cscheide\ArticleExtractor\ArticleExtractor;
use PHPUnit\Framework\TestCase;

class ExtractorTest extends TestCase {

	private $problem_sites = [
		'https://www.businesswire.com/news/home/20220607006074/en',
		'https://news.google.com/__i/rss/rd/articles/CBMihgFodHRwczovL29oaW9jYXBpdGFsam91cm5hbC5jb20vMjAyMS8wNy8yMS9zdHVkeS1jZW50cmFsLW9oaW8taG91c2luZy1zdGFiaWxpemF0aW9uLXByb2plY3QtaW1wcm92ZWQtaGVhbHRoLW91dGNvbWVzLWZvci13b21lbi1pbmZhbnRzL9IBAA?oc=5',
		'https://www.google.com/url?rct=j&sa=t&url=https://medium.com/pcmag-access/majority-of-iot-traffic-on-corporate-networks-is-insecure-report-finds-fca274d7409a%3Fsource%3Dtopic_page---------0------------------1&ct=ga&cd=CAIyGjUwNDYzZmZmOTBiY2RhNTM6Y29tOmVuOlVT&usg=AFQjCNHdAURGnK0sMRnSgOckblLTRgxUAQ',
		'https://slashdot.org/story/18/07/19/2050244/microsofts-plan-to-try-to-win-back-consumers-with-modern-life-services',
		'http://feedproxy.google.com/~r/businessinsider/~3/EChmgXESt_4/wells-fargo-close-settlement-end-probes-sales-practices-federal-prosecutors-2020-2-1028927535', // Issue #26 multiple redirects occur when browser user-agent not set
		'http://www.businesswire.com/news/home/20200213005014/en/ID-Solutions-S.r.l.-Murata-ID-Solutions-S.r.l./?feedref=JjAwJuNHiystnCoBq_hl-fLcmYSZsqlD_XPbplM8Ta6D8R-QU5o2AvY8bhI9uvWSD8DYIYv4TIC1g1u0AKcacnnViVjtb72bOP4-4nHK5iej_DoWrIhfD31cAxcB60aE', // Redirect detection issue
		'https://www.bbc.co.uk/news/uk-politics-47379565', // Issue #23 301 redirects to incomplete URL
		'https://www.fastcompany.com/3067246/innovation-agents/the-unexpected-design-challenge-behind-slacks-new-threaded-conversations',
		'http://www.mckinsey.com/industries/financial-services/our-insights/engaging-customers-the-evolution-of-asia-pacific-digital-banking?cid=other-eml-alt-mip-mck-oth-1701',
		'http://www.nhregister.com/opinion/20170116/poor-elijahs-almanack-some-choice-observations',
		'http://www.mckinsey.com/business-functions/digital-mckinsey/our-insights/the-next-generation-operating-model-for-the-digital-world?cid=reinventing-eml-alt-mip-mck-oth-1703',
		'https://t.co/kwb19AGfxl',	// Redirects
		'http://www.slate.com/articles/news_and_politics/war_stories/2017/01/trump_talks_about_himself_complains_about_media_at_first_official_event.html',
		'http://calnewport.com/blog/2017/03/13/yuval-harari-works-less-than-you/?utm_source=pocket&utm_medium=email&utm_campaign=pockethits',
		'http://www.dn.pt/sociedade/interior/ele-e-uma-estrela-pop-ele-e-o-papa-do-povo-8472352.html',  // Portuguese
		'https://hbr.org/2017/03/the-promise-of-blockchain-is-a-world-without-middlemen',				// Requires remote language check
		'http://kotaku.com/nintendo-switch-the-kotaku-review-1792776350?utm_source=pocket&utm_medium=email&utm_campaign=pockethits',
		'http://futurememes.blogspot.jp/2017/01/cognitive-easing-human-identity-crisis.html?m=1',
		'https://medium.com/personal-growth/thank-god-i-dont-have-a-job-ae2fcb6c096e',
		'https://www.cnbc.com/amp/2017/03/21/former-google-exec-shares-the-no-1-trait-great-bosses-share.html',
		'http://www.cnn.com/2017/04/07/politics/russia-us-syria-strike/index.html?adkey=bn',
		'http://gizmodo.com/how-to-survive-the-next-catastrophic-pandemic-1793487027?utm_source=pocket&utm_medium=email&utm_campaign=pockethits',
	];

	private $html_files = [
		'tests/html_test_cnbc.html'
	];

	private $overrides = [
		'https://finance.yahoo.com/news/coronavirus-stimulus-deal-broadband-internet-provisions-204533636.html',
		'https://finance.yahoo.com/news/scaramucci-skybridge-launching-bitcoin-fund-172809016.html',
	];

	private $known_problems = [

		// React problem #5 (https://github.com/crscheid/php-article-extractor/issues/5)
		'https://www.bitcoinunlimited.info/faq',

		// HTML doesn't actually contain the story without first loading some client-side javascript, similar to #5 above
		'http://mw.nikkei.com/sp/#!/article/DGXLASJC20H12_Q7A420C1000000',

		// Temporary redirect 307 to terms of service violation which prevents link from resolving
		'https://www.bloomberg.com/news/articles/2018-07-12/jpmorgan-wells-fargo-may-go-back-to-basics-with-loans-in-focus',

		// Iframes
		'http://feeds.bizjournals.com/~r/bizj_washington/~3/043koKcU8Zk/walter-reed-project-signs-day-care-preschool.html',

	];

	public function testOverrides()
	{
		$testUserAgent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.116 Safari/537.36";

		echo "\n";

		foreach($this->overrides as $url) {
			$parser = new ArticleExtractor(getenv('DETECT_LANGUAGE_KEY'), $testUserAgent, "goosecustom");
			echo "Testing Overrides: " . $url . "\n";

			$result = $parser->processURL($url);
			$this->assertNotEmpty($result['title']);
			$this->assertNotEmpty($result['text']);
			$this->assertNotEmpty($result['language']);
			$this->assertNotEmpty($result['parse_method']);
			$this->assertNotEmpty($result['language_method']);
		}
	}

	public function testProblemSites()
	{

		$testUserAgent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.116 Safari/537.36";

		echo "\n";

		foreach($this->problem_sites as $url) {
			$parser = new ArticleExtractor(getenv('DETECT_LANGUAGE_KEY'), $testUserAgent);
			echo "Testing Sites: " . $url . "\n";

			$result = $parser->processURL($url);
			$this->assertNotEmpty($result['title']);
			$this->assertNotEmpty($result['text']);
			$this->assertNotEmpty($result['language']);
			$this->assertNotEmpty($result['parse_method']);
			$this->assertNotEmpty($result['language_method']);
		}
	}

	public function testHTML() {

		echo "\n";

		foreach($this->html_files as $file) {

			echo "Testing HTML: " . $file . "\n";
			$html = file_get_contents($file);

			$parser = new ArticleExtractor(getenv('DETECT_LANGUAGE_KEY'), $testUserAgent);
			$result = $parser->processHTML($html);
			$this->assertNotEmpty($result['title']);
			$this->assertNotEmpty($result['text']);
			$this->assertNotEmpty($result['language']);
			$this->assertNotEmpty($result['parse_method']);
			$this->assertNotEmpty($result['language_method']);
		}

	}



}


tests/phpunit-bootstrap.php
<?php

require_once __DIR__ . '/../vendor/autoload.php';

date_default_timezone_set("UTC"); 

?>

app.php
<?php

require __DIR__ . '/vendor/autoload.php';


use Goose\Client as GooseClient;
use Monolog\Logger;
use Monolog\Handler\StreamHandler;
use PHPHtmlParser\Dom;

// McKinsey - very poorly formed HTML
// Slate - article had text that had more in the paragraph than others


// TODO: Check out why the raw doc works for smithsonian and mckinsey, but not the cleaned doc


// Setup the logger
date_default_timezone_set('UTC');
$logger = new Logger('parser-harness');
$logger->pushHandler(new \Monolog\Handler\ErrorLogHandler());


$test_urls = [
// 'http://www.slate.com/blogs/browbeat/2017/01/18/will_and_grace_is_returning_to_nbc_for_a_10_episode_revival.html',
// 'http://www.nhregister.com/opinion/20170116/poor-elijahs-almanack-some-choice-observations',
// 'http://www.slate.com/blogs/the_slatest/2017/01/22/audit_was_an_excuse_and_trump_is_never_releasing_tax_returns.html',
// 'http://www.slate.com/articles/news_and_politics/war_stories/2017/01/trump_talks_about_himself_complains_about_media_at_first_official_event.html',
 'http://www.mckinsey.com/industries/financial-services/our-insights/engaging-customers-the-evolution-of-asia-pacific-digital-banking?cid=other-eml-alt-mip-mck-oth-1701',
// 'https://www.fastcompany.com/3067246/innovation-agents/the-unexpected-design-challenge-behind-slacks-new-threaded-conversations',
// 'http://www.smithsonianmag.com/science-nature/day-nimbus-weather-satellie-180961686/?utm_source=smithsoniansciandnat&utm_medium=email&utm_campaign=2016701-science&spMailingID=27585738&spUserID=NzQwNDU3NTcxNDES1&spJobID=963439645&spReportId=OTYzNDM5NjQ1S0',
];


foreach($test_urls as $url) {

	log_debug("Checking: " . $url);

	// Try to get the article
	$goose = new GooseClient(['image_fetch_best' => false]);
	$article = $goose->extractContent($test_urls[0]);
	
	if ($article->getCleanedArticleText() == null) {
	
		// Get the HTML from goose
		$html_string = $article->getRawHtml();
	
		// Ok then try it a different way
		$dom = new Dom;
		$dom->load($html_string, ['whitespaceTextNode' => false]);
		
		// First, just completely remove the items we don't even care about		
		$scriptNodes = $dom->find('script, style, header, footer, input, form, button, aside, meta, link');
		foreach($scriptNodes as $node) {
			$node->delete();
			unset($node);
		}		

		$best_div = null;	

		$html = $dom->outerHtml;
		var_dump($html);		

		// Get a list of qualifying nodes we want to evaluate as the top node for content
		$contentList = buildAllNodeList($dom->root);

		// Find a target best element
		foreach($contentList as $node) {

            // Calculate the wordcount, whitecount, and ratio for this element after stripping out all tags
            $this_element_wc = str_word_count($node->text(true));
            $this_element_whitecount = substr_count($node->text(true), ' ');
            $this_element_wc_ratio = -1;
            
            if ($this_element_wc != 0) {
                $this_element_wc_ratio = $this_element_whitecount / $this_element_wc;
            }

            // Calculate the word count contribution for all children div's
            $children_wc = 0;
            $children_num = 0;
			foreach($node->getChildren() as $child) {
				if ($child->tag->name() == 'body' || $child->tag->name() == 'form' || $child->tag->name() == 'main' || $child->tag->name() == 'div' || $child->tag->name() == 'ul' || $child->tag->name() == 'li' || $child->tag->name() == 'table' || $child->tag->name() == 'span' || $child->tag->name() == 'section' || $child->tag->name() == 'article') {
					$children_num++;
					$children_wc += str_word_count($child->text(true));
				}
            }

            // This is the contribution for this particular div not including the children types above
            $my_wc_contribution = $this_element_wc - $children_wc;

            log_debug("Element:\t". $node->tag->name() . "\tTotal WC:\t" . $this_element_wc . "\tTotal White:\t" . $this_element_whitecount . "\tRatio:\t" . number_format($this_element_wc_ratio,2) . "\tElement WC:\t" . $my_wc_contribution . "\tChildren WC:\t" . $children_wc . "\tChild Contributors:\t" . $children_num . "\tBest WC:\t" . $best_div_wc . "\tBest Ratio:\t" . number_format($best_div_wc_ratio,2) . " " . $node->getAttribute('class') . "\n");

			if (strpos($node->getAttribute('class'), 'brightcovevideosingle') !== false) {
	            
	            foreach($node->getChildren() as $child) {
	            	$this_element_wc2 = str_word_count($child->text(true));
		            log_debug("Child:\t". $child->tag->name() . " Class: " . $node->getAttribute('class') . "\tTotal WC:\t" . $this_element_wc2 . "\n");
	            }
			}

            // Now check to see if this element appears better than any previous one
            
            // We do this by first checking to see if this elements WC contribution is greater than the previous
            if  ($my_wc_contribution > $best_div_wc) {
                
                // There are three conditions required for this candidate not to be chosen
                //      1. The previous best cannot be zero
                //      2. The new best is less than 5%
                //      3. The new element wc ratio is greater than the existing best element's ratio
                if (    $best_div_wc != 0 && 
                        ($my_wc_contribution / $best_div_wc) < 1.10 &&
                        $best_div_wc_ratio < $this_element_wc_ratio
                        ) {

                    log_debug("\tNot new best element - less than 10% improvement and not better wc ratio\n");
                    
                }
                // If it the previous was zero, then go ahead and update the best
                else {
                    $best_div_wc = $my_wc_contribution;
                    $best_div_wc_ratio = $this_element_wc_ratio;
                    $best_div = $node;
                    log_debug("\t *** New best element ***\n");
                }
            }
		}


		if ($best_div) {
			$text = html_entity_decode($best_div->text(true));
			log_info("Article Text Custom");	
		}
		else {
			log_info("Article Text Custom: NULL");	
		}


	}
	else {
		$text = $article->getCleanedArticleText();
		log_info("Article Text Parser");	
	}
}

// Convert items to UTF-8
$clean_utf_title = iconv(mb_detect_encoding($article->getTitle(), mb_detect_order(), true), "UTF-8", $article->getTitle());
$clean_utf_text = iconv(mb_detect_encoding($text, mb_detect_order(), true), "UTF-8", $text);

log_debug("Text: " . $clean_utf_text);



function buildAllNodeList($element) {

	$return_array = array();

	if($element->getTag()->name() != "text") {

		// Look at each child div element
		if ($element->hasChildren()) {

			foreach($element->getChildren() as $child)
			{
				// Push the children's children
				$return_array = array_merge($return_array, array_values(buildAllNodeList($child)));
	
				// Include the following tags in the counts for children and number of words
				if ($child->tag->name() == 'body' || $child->tag->name() == 'form' || $child->tag->name() == 'main' || $child->tag->name() == 'div' || $child->tag->name() == 'ul' || $child->tag->name() == 'li' || $child->tag->name() == 'table' || $child->tag->name() == 'span' || $child->tag->name() == 'section' || $child->tag->name() == 'article') {
					array_push($return_array, $child);
				}
			}
		}
	}
	return $return_array;
}

/*
 * Utility logging functions
 */
function log_debug($a) {
	global $logger;
	$logger->addDebug($a);
}
function log_info($a) {
	global $logger;
	$logger->addInfo($a);
}
function log_error($a) {
	global $logger;
	$logger->addError($a);
}
function log_warning($a) {
	global $logger;
	$logger->addWarning($a);
}

?>